src/freenet/client/filter/TheoraPacketFilter.java

```diff
@@ -1,148 +1,321 @@
+                int NEWQR = 1;
+                if (qti > 0 || pli > 0) {
+                    NEWQR = input.readBit();
+                }
+
+                int[][] NQRS = new int[2][3];
+                int[][][] QRSIZES = new int[2][3][63];
+                int[][][] QRBMIS = new int[2][3][64];
+                if (NEWQR == 0) {
+                    int qtj, plj;
+                    int RPQR = 0;
+                    if (qti > 0) {
+                        RPQR = input.readBit();
+                    }
+
+                    if (RPQR == 1) {
+                        qtj = qti - 1;
+                        plj = pli;
+                    } else {
+                        qtj = (3 * qti + pli - 1) / 3;
+                        plj = (pli + 2) % 3;
+                    }
+
+                    NQRS[qti][pli] = NQRS[qtj][plj];
+                    QRSIZES[qti][pli] = QRSIZES[qtj][plj];
+                    QRBMIS[qti][pli] = QRBMIS[qtj][plj];
+                } else {
+                    if (NEWQR != 1) {
+                        throw new UnknownContentTypeException("SETUP HEADER - NEWQR: " + NBMS + "(MUST be 0|1)");
+                    }
+
+                    int qri = 0;
+                    int qi = 0;
+
+                    QRBMIS[qti][pli][qri] = input.readInt(ilog(NBMS - 1));
+
+                    if (QRBMIS[qti][pli][qri] >= NBMS) {
+                        throw new UnknownContentTypeException("(QRBMIS[qti][pli][qri] = " + QRBMIS[qti][pli][qri] +
+                                ") >= (NBMS = " + NBMS + ") The stream is undecodable.");
+                    }
+
+                    while (true) {
+                        QRSIZES[qti][pli][qri] = input.readInt(ilog(62 - qi)) + 1;
+
+                        qi = qi + QRSIZES[qti][pli][qri];
+                        qri++;
+
+                        QRBMIS[qti][pli][qri] = input.readInt(ilog(NBMS - 1));
+
+                        if (qi < 63) {
+                            continue;
+                        } else if (qi > 63) {
+                            throw new UnknownContentTypeException("qi = " + qi + "; qi > 63 - The stream is undecodable.");
+                        }
+
+                        break;
+                    }
+
+                    NQRS[qti][pli] = qri;
+                }
+            }
+        }
+
+        int[][] HTS = new int[80][0];
+        for (int hti = 0; hti < 80; hti++) {
+            HTS[hti] = readHuffmanTable(0, HTS[hti], input);
+        }
+
+        try {
+            input.readBit();
+            Logger.minor(this, "SETUP_HEADER contains redundant bits");
+        } catch (EOFException ignored) { // should be eof
+        }
+
+        expectedPacket = Packet.FRAME;
+    }
+
+    // The minimum number of bits required to store a positive integer `a` in
+    // two’s complement notation, or 0 for a non-positive integer a.
+    private int ilog(int a) {
+        if (a <= 0) {
+            return 0;
+        }
+
+        int n = 0;
```

Comment:
It just occured to me that good old `Integer` has a `bitCount` method which does the same, probably. Maybe that could be used? (It doesn’t mention what it does with negative numbers but I assume it treats them as any other number.)